Coldcard support

humanumbrella opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

<!---
  Please describe your change(s) in detail.
  Why is this change required? What problem does it solve?
  If it fixes an open issue, please link to the issue here.
-->

## Does this PR introduce a breaking change?

<!--
  If this PR contains a breaking change,
  please also describe the impact and migration path for existing applications
-->

* [ ] Yes
* [ ] No

## Does this PR fix an open issue?

<!-- If this PR contain fixes to open issues please link them here -->

* [ ] Yes
* [ ] No

humanumbrella committed:
WIP

humanumbrella committed:
fix typo

humanumbrella committed:
add test suite link to navbar menu

humanumbrella committed:
wallet generator is braid aware, transaction reducer is braid aware, WIP

humanumbrella committed:
xpubs now check for / set XFPs ... WIP

humanumbrella committed:
basic psbt downloading ... refactor coldcard filereader (handle json and psbt)

humanumbrella committed:
getting there, WIP

humanumbrella committed:
update logo and editablename to allow clicks on the text itself

humanumbrella committed:
tweak

humanumbrella committed:
bip32 tweaks, WIP

humanumbrella committed:
some cleanup and error passback

humanumbrella committed:
signing in progress ... still need to fix multi address sigs

humanumbrella committed:
getting there. this is pretty close, more work remains

humanumbrella committed:
shoulda been in already

humanumbrella committed:
first pass of a self review

humanumbrella committed:
tweaks and make name editing better

humanumbrella committed:
save xfp on direct/indirect, remove derive xpub option

humanumbrella committed:
plumbing changes to allow regtest private node for p2wsh addresses

humanumbrella committed:
check network on bip32path init

humanumbrella committed:
test suite first pass (not right but working)

humanumbrella committed:
multi signatures working but ux not
updated testrun
refactor indirect/direct
walletdetailstext to a selector

humanumbrella committed:
missed a rename xfp => fingerprint

dhruvbansal commented on src/components/Wallet/DirectDeviceExtendedPublicKeyImporter.jsx:
```diff
@@ -7,19 +7,16 @@ import {
   ERROR,
```
Try `DirectExtendedPublicKeyImporter`

dhruvbansal commented on src/components/Wallet/IndirectDeviceExtendedPublicKeyImporter.jsx:
```diff
@@ -0,0 +1,263 @@
+import React from "react";
```
We need to make this component abstract and not concrete.  Add a required prop `Reader` to the component that concrete classes (e.g. `HermitExtendedPublicKeyImporter` or `ColdcardExtendedPublicKeyImporter`) can use to pass in a custom component to handle request/parse cycle.

```
// in IndirectExtendedPublicKeyImporter.jsx

class IndirectExtendedPublicKeyImporter extends React.Component {

  render = () => {
    const {Reader} = this.props;
    return (
      <div>
        <Reader 
          hasError={this.hasBIP32PathError()} 
          onReceive={this.onReceive} 
          interaction={this.interaction()} />
      </div>
    );
  }

}

// in coldcard.js

class ColdcardExtendedPublicKeyImporter extends React.Component {

  defaultBIP32Path = () => {
    const {network, addressType} = this.props;
    // Coldcard-specific default...
  }

  render = () => {
    return (
      <IndirectExtendedPublicKeyImporter
        defaultBIP32Path={this.defaultBIP32Path()}
        Reader={ColdcardFileReader}
      />
    );
  }

}


// in hermit.js

class HermitExtendedPublicKeyImporter extends React.Component {

  render = () => {
    return (
      <IndirectExtendedPublicKeyImporter
        Reader={HermitReader}
      />
    );
  }

}
```

humanumbrella replied:
will review what I did re: this more tomorrow, but I have something working for now.

dhruvbansal commented on src/components/ScriptExplorer/DirectHardwareWalletSignatureImporter.jsx:
```diff
@@ -1,6 +1,5 @@
 import React from "react";
 import PropTypes from "prop-types";
-
```
This file/component should be `DirectSignatureImporter`

dhruvbansal commented on src/components/ScriptExplorer/IndirectHardwareWalletSignatureImporter.jsx:
```diff
@@ -0,0 +1,353 @@
+import React from "react";
```
Should be named `IndirectSignatureImporter`

dhruvbansal commented on src/actions/extendedPublicKeyImporterActions.js:
```diff
@@ -37,3 +42,23 @@ export function setExtendedPublicKeyImporterVisible(value) {
   };
 }
+
+export function generateRichExtendedPublicKeys(extendedPublicKeyImporters) {
```
Move this to the `WalletGenerator` component.

dhruvbansal commented on src/components/TestSuiteRun/Seed.jsx:
```diff
@@ -17,26 +17,34 @@ class SeedBase extends React.Component {
+            </ol>
+          </Grid>
       </Grid>
+      <Grid style={{marginTop:'2em', marginBottom:'2em'}}>
```
Put a `if coldcard` guard on this

humanumbrella committed:
first round pr review edits

humanumbrella committed:
pr review revision

humanumbrella committed:
cleanup

humanumbrella committed:
working on indirect abstractions and other refactoring

humanumbrella committed:
fixes #127

humanumbrella committed:
refactoring and fixing bip32paths for coldcard xpub exporter

humanumbrella committed:
generate wallet config text locally instead of relying on stored in fixtures

humanumbrella committed:
tests check that fingerprint is included (and correct) too

humanumbrella committed:
fixes bug when importing config (unknown client) and trying to use private node
also if you test connection to private node and then import config (unknown client)
that was another bug.

humanumbrella committed:
fixes #106

humanumbrella committed:
fixes #120

humanumbrella committed:
fixes #106 continued

humanumbrella committed:
add-on to #111 to make error more obvious

humanumbrella committed:
thoughts on bug with N-ly signed multi-address spend

humanumbrella committed:
fixes #118, #119, #125

humanumbrella committed:
fix bug where cannot reload same file after an error

humanumbrella committed:
fixes #124

humanumbrella committed:
Addresses #130

humanumbrella committed:
Fixes #117

dhruvbansal commented on src/components/Hermit/HermitExtendedPublicKeyImporter.jsx:
```diff
@@ -1,8 +1,6 @@
 import React from "react";
```
Add a note why you're not using the shared indirect helper component class (no need to input BIP32 path!)

dhruvbansal commented on src/components/ScriptExplorer/SignatureImporter.jsx:
```diff
@@ -304,72 +314,159 @@ class SignatureImporter extends React.Component {
       errback("Not enough signatures (must be exactly one for each input).");
```
"Must be at least one signature for each input"

dhruvbansal commented on src/components/ScriptExplorer/SignatureImporter.jsx:
```diff
@@ -304,72 +314,159 @@ class SignatureImporter extends React.Component {
+
+    if (inputsSignatures.length % inputs.length !== 0) {
+      errback(
+        "Not the right number of signatures (must be exactly one for each input)."
```
"Number of signatures must be a multiple of number of inputs."

dhruvbansal replied:
Also add a `return` to short-circuit out of this function after calling `errback`

dhruvbansal commented on src/components/ScriptExplorer/SignatureImporter.jsx:
```diff
@@ -304,72 +314,159 @@ class SignatureImporter extends React.Component {
+      );
     }
 
+    const numSignatureSets = inputsSignatures.length / inputs.length;
```
`numSignatureSets` cannot be larger than number of required signers (also == to the number of signature importers)

dhruvbansal commented on src/components/ScriptExplorer/SignatureImporter.jsx:
```diff
@@ -304,72 +314,159 @@ class SignatureImporter extends React.Component {
+      // We land here if a PSBT has been uploaded with multiple signature sets.
+      // In case we already have some signatures, filter those out.
+      let signaturesToCheck = this.filterKnownSignatures(inputsSignatures);
+      let sigNumber = number; // so we can iterate, number is a const from props.
```
This is `signatureSetNumber`

humanumbrella committed:
Add fixture Caravan wallet configs to public directory
adds working signature parser for N-Ly signed PSBTs

humanumbrella committed:
add coldcard label/link to tested help section

humanumbrella committed:
add coldcard segwit pubkey tests
fix signatureimporter visibility on multi-signed PSBT upload

humanumbrella committed:
text-based-signature uses proper indirect pattern

humanumbrella committed:
ask for xfp with pubkey in tests

humanumbrella committed:
es linting fixes

humanumbrella committed:
expanding commentary about signatures

humanumbrella committed:
fixes

humanumbrella committed:
chore(other) package updates
